### ISSUE TITLE
Provider Review: Acceptance Tests

### ISSUE THREAD

**Issue by cgriggs01** [tencentyun/terraform-provider-tencentcloud#13]
Hi Tencent Team,

I was able to run the acceptance test with the credentials that you have provided HashiCorp for the official provider program. 

The test ran easily but there are 3 tests that are failing consistently:

```
=== RUN   TestAccTencentCloudInstance_basic
--- FAIL: TestAccTencentCloudInstance_basic (4.51s)
	testing.go:513: Step 0 error: Error applying: 1 error(s) occurred:

		* tencentcloud_instance.foo: 1 error(s) occurred:

		* tencentcloud_instance.foo: tencentcloud_instance got error, code:InvalidCloudDisk.SoldOut, message:The specified cloud disk is sold out, because `[19014]当前配置云盘已售罄`
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: queryInstancesStatus, empty instanceIds

		State: <no state>
```

```
=== RUN   TestAccTencentCloudInstance_imageIdChanged
--- FAIL: TestAccTencentCloudInstance_imageIdChanged (4.27s)
	testing.go:513: Step 0 error: Error applying: 1 error(s) occurred:

		* tencentcloud_instance.hello: 1 error(s) occurred:

		* tencentcloud_instance.hello: tencentcloud_instance got error, code:InvalidCloudDisk.SoldOut, message:The specified cloud disk is sold out, because `[19014]当前配置云盘已售罄`
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Check failed: queryInstancesStatus, empty instanceIds

		State: <no state>
```

 ```
=== RUN   TestAccTencentCloudInstance_PrePaid
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
ins-nwl1dm0a
--- FAIL: TestAccTencentCloudInstance_PrePaid (95.96s)
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Error applying: 1 error(s) occurred:

		* tencentcloud_instance.prepaid (destroy): 1 error(s) occurred:

		* tencentcloud_instance.prepaid: prepaid instance ins-nwl1dm0a is not allowed to be deleted in terraform

		State: data.tencentcloud_availability_zones.my_favorate_zones:
		  ID = 340736911
		  provider = provider.tencentcloud
		  zones.# = 3
		  zones.0.description = 广州二区
		  zones.0.id = 100002
		  zones.0.name = ap-guangzhou-2
		  zones.0.state = AVAILABLE
		  zones.1.description = 广州三区
		  zones.1.id = 100003
		  zones.1.name = ap-guangzhou-3
		  zones.1.state = AVAILABLE
		  zones.2.description = 广州四区
		  zones.2.id = 100004
		  zones.2.name = ap-guangzhou-4
		  zones.2.state = AVAILABLE
		data.tencentcloud_image.my_favorate_image:
		  ID = 370168848
		  provider = provider.tencentcloud
		  filter.# = 1
		  filter.2430114294.name = image-type
		  filter.2430114294.values.# = 1
		  filter.2430114294.values.0 = PUBLIC_IMAGE
		  image_id = img-dkwyg6sr
		  os_name = centos
		data.tencentcloud_instance_types.my_favorate_instance_types:
		  ID = 616936589
		  provider = provider.tencentcloud
		  cpu_core_count = 1
		  filter.# = 1
		  filter.1670263193.name = instance-family
		  filter.1670263193.values.# = 1
		  filter.1670263193.values.0 = S1
		  instance_types.# = 2
		  instance_types.0.availability_zone = ap-guangzhou-2
		  instance_types.0.cpu_core_count = 1
		  instance_types.0.family = S1
		  instance_types.0.instance_type = S1.SMALL2
		  instance_types.0.memory_size = 2
		  instance_types.1.availability_zone = ap-guangzhou-3
		  instance_types.1.cpu_core_count = 1
		  instance_types.1.family = S1
		  instance_types.1.instance_type = S1.SMALL2
		  instance_types.1.memory_size = 2
		  memory_size = 2
		tencentcloud_instance.prepaid:
		  ID = ins-nwl1dm0a
		  provider = provider.tencentcloud
		  allocate_public_ip = false
		  availability_zone = ap-guangzhou-2
		  data_disks.# = 0
		  image_id = img-dkwyg6sr
		  instance_charge_type = PREPAID
		  instance_charge_type_prepaid_period = 1
		  instance_name = terraform_automation_test_kuruk_prepay
		  instance_status = RUNNING
		  instance_type = S1.SMALL2
		  private_ip = 172.16.0.7
		  subnet_id = subnet-0iky1dqy
		  system_disk_size = 50
		  system_disk_type = LOCAL_BASIC
		  vpc_id = vpc-kg60ct5z

		  Dependencies:
		    data.tencentcloud_availability_zones.my_favorate_zones
		    data.tencentcloud_image.my_favorate_image
		    data.tencentcloud_instance_types.my_favorate_instance_types
```

Any thoughts on what might be causing these error? We will need to get all the acceptance tests passing before we can officially release the provider.

Thanks,
Chris

**Comment by zqfan** [comment]
Hi Chris,
The first two are caused by same reason, the data disk specified in acceptance test is unfortunately sold out, I think if we adjust the default availability zone value in acc test, then this problem will be fixed.

The third one is caused by a feature of Tencent Cloud API, the prepaid instance/disk cannot be deleted manually, only can be expired and deleted automatically. Do you have any idea to handle such case in Terraform? For now, we just fail the delete operation.

Thanks,
ZhiQiang

**Comment by zqfan** [comment]
@cgriggs01 I've submitted a fix for these three case : https://github.com/tencentyun/terraform-provider-tencentcloud/pull/17

**Comment by cgriggs01** [comment]
Thanks for submitting those @zqfan.

After looking through the source test code, it appears that only some of the resources and data sources have acceptance tests written for them. Each resource/data source should have a corresponding tests along with it.  

Once there is test coverage over the entire provider, we can start the technical review.

**Comment by cgriggs01** [comment]
Hi @zqfan,

Any progress on the rest of the acceptance tests?

Best,
Chris.

**Comment by zqfan** [comment]
Hi @cgriggs01 ,

The past two weeks are Chinese's Luna New Year Festival, so the progress was pending during that time. We are still working on it. Currently, acceptance tests for VPC data sources and resources
are under review.

Thanks for your attention. Cheers!
ZhiQiang.

**Comment by cgriggs01** [comment]
Any update on the progress of the acceptance tests?

Thanks,
Chris

**Comment by zqfan** [comment]
@cgriggs01 
We have added vpc & cbs (block storage) acceptance as well, and another development iteration including new resources/data sources are merged with acceptance.

However, we still have introduced new code which dependent resource cannot be created yet, which makes the acceptance cannot 100% cover, particularly the alb_server_attachment resource and eip_association with network interface resource

is it possible to leave just one or two resource not covered by acceptance test case?

Thanks,
Zhiqiang

**Comment by cgriggs01** [comment]
I am having issues building the provider now, do you have any ideas what may be causing this?

```
~/Repository/src/github.com/tencentyun/terraform-provider-tencentcloud (dev ✘)✹✭ ᐅ make build
==> Checking that code complies with gofmt requirements...
go install
# github.com/tencentyun/terraform-provider-tencentcloud/tencentcloud
tencentcloud/resource_tc_dnat.go:69:10: undefined: vpc.NewAddDnaptRuleRequest
tencentcloud/resource_tc_dnat.go:80:23: conn.AddDnaptRule undefined (type *vpc.Client has no field or method AddDnaptRule)
tencentcloud/resource_tc_dnat.go:133:10: undefined: vpc.NewDeleteDnaptRuleRequest
tencentcloud/resource_tc_dnat.go:136:21: undefined: vpc.DnaptRuleInput
tencentcloud/resource_tc_dnat.go:145:23: conn.DeleteDnaptRule undefined (type *vpc.Client has no field or method DeleteDnaptRule)
tencentcloud/resource_tc_dnat.go:155:25: undefined: vpc.AddDnaptRuleRequest
tencentcloud/resource_tc_dnat.go:163:42: undefined: vpc.DnaptRule
tencentcloud/resource_tc_dnat.go:173:11: undefined: vpc.DnaptRule
tencentcloud/resource_tc_nat_gateway.go:89:72: response.NatGatewayId undefined (type *vpc.CreateNatGatewayResponse has no field or method NatGatewayId)
tencentcloud/service_vpc.go:58:81: undefined: vpc.DnaptRule
tencentcloud/resource_tc_nat_gateway.go:89:72: too many errors
make: *** [build] Error 2
```

**Comment by zqfan** [comment]
@cgriggs01 We're not tracking all vendor's code in our project but only have the vendor.json file, so you might need to use govendor sync -v to synchronize latest vendor's code

**Comment by zqfan** [comment]
@cgriggs01 
Hi Chris,

We resolved the acceptance issue in dev branch, and will merge to master branch soon.

For the vendor directory problem, we didn't track them for a bit of legal concern even though we published our source code under MPL license. But as you says, the aws provider has done it from early stage, I think it is acceptable for us to track those code as well.

**Comment by zqfan** [comment]
@cgriggs01 
The fix in dev branch has been merged into master branch, along with new container resources, and released as v1.2.0

**Comment by zqfan** [comment]
@cgriggs01 
Hi Chris,

I noticed that from April 5 to April 6, there are some vpc, nat gateway resources are created but not released, which causes quota limit problem when I run acceptance test this morning. I think it is not some kind of resource leak problem because previously I run acceptance test and they are cleaned as expected. So I'm wondering if Terraform Provider Development Process will use the credential to run some test from .tf file which cause the resources will not be destroyed automatically?

Best Regards,
ZhiQiang

**Comment by cgriggs01** [comment]
@zqfan 

The acceptance test are always run from `make testacc`. That is strange the the resources were not deleted. 

The first time I ran the tests last week, I believe one test failed, then running them again that day, I started to hit the quota issues.

I will test again and let you know. Thanks for your help!

Best,
Chris

**Comment by cgriggs01** [comment]
For the accteptance testing, Terraform providers should all use our standard [GNUmakefile](https://github.com/terraform-providers/terraform-provider-scaffolding/blob/master/GNUmakefile) and directory of four corresponding [bash scripts](https://github.com/terraform-providers/terraform-provider-scaffolding/tree/master/scripts). Two of them are for testing and two are for the release process. 

Can you please add these to the provider repository. This is how we will be running the acceptance tests with `make testacc`. 

Thanks,
Chris

**Comment by zqfan** [comment]
@cgriggs01 

should I just **copy** those files you've mentioned to this repo? I will add another pr for it and add you to the reviewer list.

 I noticed that [aws make file](https://github.com/terraform-providers/terraform-provider-aws/blob/master/GNUmakefile) is slightly different to this one https://github.com/terraform-providers/terraform-provider-scaffolding/blob/master/GNUmakefile , so should I do some customization for this repo also?

**Comment by zqfan** [comment]
@cgriggs01 
I have create a pr https://github.com/tencentyun/terraform-provider-tencentcloud/pull/51 to resolve the issue, I have checked to run make test and make testacc in local environment, please help us to review it.

For now it is in dev branch, once it is merged into dev branch, then it will be released in master branch.

**Comment by zqfan** [comment]
@cgriggs01 
#51 has been released into master branch. I just checked `make test` and `make vet`, both seems fine.

**Comment by zqfan** [comment]
@cgriggs01 

Can we move to next step now? Or is there anything else we should fix in this step?

**Comment by zqfan** [comment]
@cgriggs01 
Hi Chris, is there any update of the review progress?
Thanks!

**Comment by cgriggs01** [comment]
@zqfan 

I've run the acceptance test numerous times and I've continuously failed the same two test:
TestAccTencentCloudContainerClusterInstance_basic
TestAccTencentCloudDnat_basic

Included here are the [standard output file](https://github.com/tencentyun/terraform-provider-tencentcloud/files/1961246/standard-acctest.txt) as well as the overly detailed [debug output file](https://github.com/tencentyun/terraform-provider-tencentcloud/files/1961245/debug-acctest.txt) for your reference. 

Once I can run the tests locally and they all pass, your provider can move on to the next phase of the program which is a in-depth code review from the Terraform team. 

Best,
Chris

**Comment by zqfan** [comment]
@cgriggs01 

The acceptance test case fails not because test code itself but due to Cloud API has been made some incompatible changes. 
For the dnat test case, it can be resolved by upgrade cloud sdk, #54 will handle it. 
For the container cluster instance test case, we've already requested the container service team to track this issue. It might take some time for it.

Thanks.

**Comment by cgriggs01** [comment]
@zqfan 

Thanks for the explanation regarding those tests. 

I will close this issue now and a Terraform engineer will open a new issue with feedback once they have reviewed the provider. 

Thank you for your patience throughout this process. 

Best,
Chris

**Comment by zqfan** [comment]
@cgriggs01 
Hi Chris,

The problem in acceptance has been fixed, I run make testacc this morning and it runs ok.

I didn't see any new issue opened for the code review process, who should I contact?

Best Regards,
ZhiQiang
